File: 94-binary-tree-inorder-traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<TreeNode *> s;
        TreeNode * p = root;
        
        while ( !s.empty() || p ) {
            if ( p ) {
                s.push(p);
                p = p->left;
            } else {
                TreeNode * n = s.top(); s.pop();
                result.push_back(n->val);
                p = n->right;
            }
        }
        return result;
    }
};

File: 94-binary-tree-inorder-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res, stack = [], []
        curr = root
        while (curr or len(stack)):
            while curr:
                stack.append(curr)
                curr = curr.left
            
            curr = stack.pop(-1)
            res.append(curr.val)
            curr = curr.right        
        return res
